File: inverted-index/src/main/java/ru/itmo/sparsifiermodel/query/FieldValueAsScoreQuery.java
package ru.itmo.sparsifiermodel.query;

import org.apache.lucene.index.DocValues;
import org.apache.lucene.index.LeafReaderContext;
import org.apache.lucene.index.NumericDocValues;
import org.apache.lucene.search.*;

import java.io.IOException;
import java.util.Objects;

/**
 * Java is used so that this class can be used with PyLucene
 */
public class FieldValueAsScoreQuery extends Query {

    private final String field;
    private final float queryTermValue;

    public FieldValueAsScoreQuery(String field, float queryTermValue) {
        this.field = Objects.requireNonNull(field);
        if (Float.isInfinite(queryTermValue) || Float.isNaN(queryTermValue)) {
            throw new IllegalArgumentException("Query term value must be finite and non-NaN");
        }
        this.queryTermValue = queryTermValue;
    }

    @Override
    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) {
        return new Weight(this) {
            @Override
            public boolean isCacheable(LeafReaderContext ctx) {
                return DocValues.isCacheable(ctx, field);
            }

            @Override
            public Explanation explain(LeafReaderContext context, int doc) {
                throw new UnsupportedOperationException();
            }

            @Override
            public Scorer scorer(LeafReaderContext context) throws IOException {
                return new Scorer(this) {

                    private final NumericDocValues iterator = context.reader().getNumericDocValues(field);

                    @Override
                    public float score() throws IOException {
                        final int docId = docID();
                        assert docId != DocIdSetIterator.NO_MORE_DOCS;
                        assert iterator.advanceExact(docId);
                        return Float.intBitsToFloat((int) iterator.longValue()) * queryTermValue * boost;
                    }

                    @Override
                    public int docID() {
                        return iterator.docID();
                    }

                    @Override
                    public DocIdSetIterator iterator() {
                        return iterator;
                    }

                    @Override
                    public float getMaxScore(int upTo) {
                        return Float.MAX_VALUE;
                    }
                };
            }
        };
    }

    @Override
    public String toString(String field) {
        StringBuilder builder = new StringBuilder();
        builder.append("FieldValueAsScoreQuery [field=");
        builder.append(this.field);
        builder.append(", queryTermValue=");
        builder.append(this.queryTermValue);
        builder.append("]");
        return builder.toString();
    }

    @Override
    public void visit(QueryVisitor visitor) {
        if (visitor.acceptField(field)) {
            visitor.visitLeaf(this);
        }
    }

    @Override
    public boolean equals(Object other) {
        return sameClassAs(other) && equalsTo(getClass().cast(other));
    }

    private boolean equalsTo(FieldValueAsScoreQuery other) {
        return field.equals(other.field)
                && Float.floatToIntBits(queryTermValue) == Float.floatToIntBits(other.queryTermValue);
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int hash = classHash();
        hash = prime * hash + field.hashCode();
        hash = prime * hash + Float.floatToIntBits(queryTermValue);
        return hash;
    }
}

File: inverted-index/src/main/kotlin/ru/itmo/sparsifiermodel/query/FieldValueAsScoreQuery.kt
package ru.itmo.sparsifiermodel.query

import org.apache.lucene.index.LeafReaderContext
import org.apache.lucene.search.DocIdSetIterator
import org.apache.lucene.search.Explanation
import org.apache.lucene.search.IndexSearcher
import org.apache.lucene.search.Query
import org.apache.lucene.search.QueryVisitor
import org.apache.lucene.search.ScoreMode
import org.apache.lucene.search.Scorer
import org.apache.lucene.search.Weight

class FieldValueAsScoreQuery(private val field: String) : Query() {

    override fun createWeight(searcher: IndexSearcher, scoreMode: ScoreMode, boost: Float): Weight =
        object : Weight(this) {
            override fun isCacheable(ctx: LeafReaderContext?): Boolean {
                TODO("Not yet implemented")
            }

            override fun explain(context: LeafReaderContext?, doc: Int): Explanation {
                TODO("Not yet implemented")
            }

            override fun scorer(context: LeafReaderContext): Scorer = object : Scorer(this) {
                val iterator = context.reader().getNumericDocValues(field)

                override fun score(): Float {
                    val docId = docID()
                    require(docId != DocIdSetIterator.NO_MORE_DOCS)
                    require(iterator.advanceExact(docId))
                    return Float.fromBits(iterator.longValue().toInt()) * boost
                }

                override fun docID(): Int = iterator.docID()
                override fun iterator(): DocIdSetIterator = iterator
                override fun getMaxScore(upTo: Int): Float = Float.MAX_VALUE
            }
        }

    override fun equals(other: Any?): Boolean {
        TODO("Not yet implemented")
    }


    override fun hashCode(): Int {
        val prime = 31
        var hash = classHash()
        hash = prime * hash + field.hashCode()
        return hash
    }

    override fun toString(field: String?): String = buildString {
        append("FieldValueAsScoreQuery [field=")
        append(this@FieldValueAsScoreQuery.field)
        append("]")
    }

    override fun visit(visitor: QueryVisitor) {
        if (visitor.acceptField(field)) {
            visitor.visitLeaf(this)
        }
    }
}
